web/src/notifications.js

```diff
@@ -253,33 +229,112 @@ export function process_notification(notification) {
         content += " [...]";
     }
 
-    if (notice_memory.has(key)) {
-        msg_count = notice_memory.get(key).msg_count + 1;
-        title = msg_count + " messages from " + title;
-        notification_object = notice_memory.get(key).obj;
-        notification_object.close();
+    return content;
+}
+
+function debug_notification_source_value(message) {
+    let notification_source;
+
+    if (message.type === "private" || message.type === "test-notification") {
+        notification_source = "pm";
+    } else if (message.mentioned) {
+        notification_source = "mention";
+    } else if (message.alerted) {
+        notification_source = "alert";
+    } else {
+        notification_source = "stream";
+    }
+
+    blueslip.debug("Desktop notification from source " + notification_source);
+}
+
+function get_notification_key(message) {
+    let key;
+
+    if (message.type === "private" || message.type === "test-notification") {
+        key = message.display_reply_to;
+    } else {
+        key = message.sender_full_name + " to " + message.stream + " > " + message.topic;
```

Comment:
I guess we don't need to mark it for translation because the key is not visible to the user.

Reply:
Correct.

Reply:
Though we might want to think about rewriting this to use some more reasonable separators between fields or some other serialization that is robust to say a user whose name is set to something like `Tim Abbott to design > notifications`

Reply:
Didn't quite grasp your idea here. Are you suggesting that it wouldn't be safe if the user's name is set to `Tim Abbott to design > notifications`?

Reply:
Yeah, which is not a realistic case; but the point is that with a data structure where objects are meant to be unique conversations, we should design the way we construct keys such that collisions are impossible. 

I think just adding `direct:` and `stream:` would be insufficient, but I think we can leave for this for a follow-up issue: https://github.com/zulip/zulip/issues/26729